src/orion/core/evc/tree.py

```diff
@@ -0,0 +1,417 @@
+            Note that the callable object is expected to return an object which will be set as the
+            current node's item (in the resulting tree), and the parent node or the children nodes
+            depending on the direction of function application.
+        node: None, `orion.core.evc.TreeNode` or list
+            Can be either
+            `None`: function is applied on current node only
+            `self.parent`: function is applied recursively climbing up the tree until the root
+            `self.children`: function is applied recursively going down the tree until the leafs
+
+        Examples
+        --------
+        .. code-block:: python
+            :linenos:
+
+            # Tree structure
+            # a
+            # |   \  \   \
+            # b    c  d   e
+            # | \         |
+            # f  g        h
+            #
+
+            root = TreeNode(1)
+            b = TreeNode(1, root)
+            TreeNode(1, root)
+            TreeNode(1, root)
+            e = TreeNode(1, root)
+
+            f = TreeNode(1, b)
+            TreeNode(1, b)
+
+            h = TreeNode(1, e)
+
+            def increment(node, children):
+                for child in TreeNode(0, None, children):
+                    child.item += 1
+
+                return node.item + 1, children
+
+            # Should return
+            #
+            # 2
+            # |   \  \   \
+            # 3    3  3   3
+            # | \         |
+            # 4  4        4
+
+            rval = root.map(increment, root.children)
+            assert [node.item for node in rval] == [2, 3, 4, 4, 3, 3, 3, 4]
+
+            def increment_parent(node, parent):
+                if parent is not None:
+                    for parent in parent.root:
+                        parent.item += 1
+
+                return node.item + 1, parent
+
+            # Should return
+            #
+            # 4
+            # |
+            # 3
+            # |
+            # 2
+
+            rval = f.map(increment_parent, f.parent)
+            assert [node.item for node in rval.root] == [4, 3, 2]
+
+            rval = h.map(increment_parent, h.parent)
+            assert [node.item for node in rval.root] == [4, 3, 2]
+
+        """
+        if node is None:
+            rval, _ = function(self, None)
+            return TreeNode(rval)
+        elif node is self.parent:
+            rval_parent_node = node.map(function, node.parent)
+            rval, parent_node = function(self, rval_parent_node)
+            return TreeNode(rval, parent_node)
+        elif node is self.children:
+            rval_children_nodes = [child.map(function, child.children) for child in self.children]
+            rval, children_nodes = function(self, rval_children_nodes)
+            return TreeNode(rval, parent=None, children=children_nodes)
+        else:
+            raise TypeError("Invalid nodes: %s" % str(node))
```

Comment:
To continue comment: https://github.com/mila-udem/orion/pull/84#discussion_r191058243

What do you mean if I agree with this one?

Reply:
@tsirif Ping :)